phonenumberfmt/format_phone_number_list.py

```diff
@@ -0,0 +1,161 @@
+        - Attempt validation by removing first digit, and then
+          adding the dialing code.
+          (This helps if we have a leading zero which implies
+           the current country's dialing prefix)
+        - Attempt validation by removing first two digits and then
+          adding the dialing code.
+          (This helps if we have two leading zeroes which
+          implies the current country's dialing prefix)
+
+    Under the hood, this uses Google's libphonenumber. If this
+    method doesn't find a phone number valid, then it must really
+    not be a valid number.
+    """
+    phone_number = str(phone_number)
+
+    if not phone_number or len(phone_number) < 3:
+        return []
+
+    dialing_prefix = dialing_prefix_for_region(implied_phone_region) or ""
+
+    valid_phone_numbers = collections.OrderedDict()
+    for number in [
+        phone_number,
+        *re.split(r"/|,", phone_number),
+        *phone_number.split(" "),
+    ]:
+        try:
+            unquoted_number = unquote(number)
+        except Exception:
+            unquoted_number = number
+        valid_numbers = []
+        _append_valid_number(
+            valid_numbers,
+            unquoted_number.split("#", maxsplit=1)[0],
+            dialing_prefix,
+            fmt,
+        )
+        _append_valid_number(
+            valid_numbers,
+            unquoted_number.split(":", maxsplit=1)[0],
+            dialing_prefix,
+            fmt,
+        )
+
+        cleaned_number = "".join(re.findall("[0-9+]+", unquoted_number))
+        _append_valid_number(valid_numbers, cleaned_number, dialing_prefix, fmt)
+        _append_valid_number(
+            valid_numbers, "+%s" % cleaned_number, dialing_prefix, fmt
+        )
+        if cleaned_number.startswith("00"):
+            _append_valid_number(
+                valid_numbers,
+                "+%s" % cleaned_number.replace("00", "", 1),
+                dialing_prefix,
+                fmt,
+            )
+        if len(valid_numbers) == 0 and not _is_already_contained(
+            valid_phone_numbers, cleaned_number
+        ):
+            _append_valid_number(
+                valid_numbers,
+                "%s%s" % (dialing_prefix, cleaned_number),
+                dialing_prefix,
+                fmt,
+            )
+            if len(valid_numbers) == 0:
+                _append_valid_number(
+                    valid_numbers,
+                    "%s%s" % (dialing_prefix, cleaned_number[1:]),
+                    dialing_prefix,
+                    fmt,
+                )
+            if len(valid_numbers) == 0:
+                _append_valid_number(
+                    valid_numbers,
+                    "%s%s" % (dialing_prefix, cleaned_number[2:]),
+                    dialing_prefix,
+                    fmt,
+                )
+
+        for valid_number in valid_numbers:
+            valid_phone_numbers[valid_number] = valid_number
```

Comment:
Maybe a `set()` data structure storing the valid phone numbers would help here.

Reply:
I need an ordered set in order to keep the order of insertions which is necessary to retrieve the first valid one

Reply:
Yes, the order might be important for the users of this library. But the internal implementation doesn't care about the order, right? So in case you go with the python generator then you return the numbers on the fly in the exact order you "find" them.

Reply:
the internal implementation in some case depends on the order, for example for this number:
"+97144220125#40", it should retrieve only "+97144220125", but it seems also this number is a valid one: "+9714422012540" (with "40" at the end) which we do not want to be added in the  "valid" number list.

And regarding to return a Generator, for each new valid number we needs to do some check in the previous valid numbers in order to add it or not. This means we need to create the entire set of values and then return from this method. Returning a generator from this already created set is redundant and ads just a layer of extra processing.